core/src/main/scala/org/bitcoins/core/wallet/utxo/UTXOSpendingInfo.scala

```diff
@@ -94,14 +118,121 @@ object ConditionalPath {
+                  witnessOpt.getOrElse(throw new IllegalArgumentException(
+                    "Script Witness must be defined for (nested) Segwit input")),
+                  conditionalPath
+                )
+              case nonWitnessSPK: RawScriptPubKey =>
+                P2SHNoNestSpendingInfoSingle(outPoint,
+                                             output.value,
+                                             p2sh,
+                                             signer,
+                                             hashType,
+                                             nonWitnessSPK,
+                                             conditionalPath)
+              case _: P2SHScriptPubKey =>
+                throw new IllegalArgumentException("Cannot have nested P2SH")
+              case _: UnassignedWitnessScriptPubKey =>
+                throw new UnsupportedOperationException(
+                  s"Unsupported ScriptPubKey ${output.scriptPubKey}")
+            }
+        }
+      case wspk: WitnessScriptPubKeyV0 =>
+        val witnessOpt = scriptWitnessOpt match {
+          case Some(witness: ScriptWitnessV0) => Some(witness)
+          case None                           => None
+          case Some(_: ScriptWitness) =>
+            throw new UnsupportedOperationException(
+              "Only v0 Segwit is currently supported")
+        }
+
+        SegwitV0NativeUTXOSpendingInfoSingle(
+          outPoint,
+          output.value,
+          wspk,
+          signer,
+          hashType,
+          witnessOpt.getOrElse(
+            throw new IllegalArgumentException(
+              "Script Witness must be defined for Segwit input")),
+          conditionalPath
+        )
+      case wspk: UnassignedWitnessScriptPubKey =>
+        UnassignedSegwitNativeUTXOSpendingInfo(
+          outPoint,
+          output.value,
+          wspk,
+          Vector(signer),
+          hashType,
+          scriptWitnessOpt.getOrElse(EmptyScriptWitness),
+          conditionalPath)
+      case rawSPK: RawScriptPubKey =>
+        RawScriptUTXOSpendingInfoSingle(outPoint,
+                                        output.value,
+                                        rawSPK,
+                                        signer,
+                                        hashType,
+                                        conditionalPath)
+    }
+  }
+}
+
+sealed trait BitcoinUTXOSpendingInfo
+    extends UTXOSpendingInfo
+    with BitcoinUTXOSpendingInfoSingle {
+
+  override def toSingle(signerIndex: Int): BitcoinUTXOSpendingInfoSingle = {
+    BitcoinUTXOSpendingInfoSingle(outPoint,
+                                  output,
+                                  signers(signerIndex),
+                                  redeemScriptOpt,
+                                  scriptWitnessOpt,
+                                  hashType,
+                                  conditionalPath)
+  }
+}
+
 object BitcoinUTXOSpendingInfo {
 
   def apply(
```

Comment:
Could this function be changed so if `signers.size == 1` then it uses `BitcoinUTXOSpendingInfoSingle.apply`

Reply:
I messed around with this idea for quite a while but in the end it turned out to be quite complicated and I think

1. It isn't really necessary since BitcoinUTXOSpendingInfo extends BitcoinUTXOSpendingInfoSingle and
2. I think that it is good to keep the two applies somewhat separate so as to make sure people get expected behavior and failing if impossible rather than trying correct for people when they make mistakes.